credentials/credentials.go

```diff
@@ -161,17 +147,18 @@ func (c *tlsCreds) ClientHandshake(addr string, rawConn net.Conn, timeout time.D
 		cfg.ServerName = addr[:colonPos]
 	}
 	conn := tls.Client(rawConn, cfg)
-	if timeout == 0 {
-		err = conn.Handshake()
-	} else {
-		go func() {
-			errChannel <- conn.Handshake()
-		}()
-		err = <-errChannel
-	}
-	if err != nil {
-		rawConn.Close()
-		return nil, nil, err
+	errChannel := make(chan error, 1)
+	go func() {
+		errChannel <- conn.Handshake()
+	}()
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			rawConn.Close()
+			return nil, nil, err
+		}
+	case <-ctx.Done():
```

Comment:
Your change makes ClientHandshake terminates when ctx is cancelled. But This makes the goroutine created in line 151 leak if conn.Handshake() stucks. It seems this does not really resolve the issue?